File: Dam-Control-System/src/main/java/com/SmartDam/Dam_Control_System/controller/DamController.java
package com.SmartDam.Dam_Control_System.controller;

import com.SmartDam.Dam_Control_System.entity.ControlLog;
import com.SmartDam.Dam_Control_System.entity.ReservoirState;
import com.SmartDam.Dam_Control_System.repository.ControlLogRepository;
import com.SmartDam.Dam_Control_System.repository.ReservoirStateRepository;
import com.SmartDam.Dam_Control_System.service.DamControlEngineService;
import com.SmartDam.Dam_Control_System.service.WeatherForecastService;
import com.SmartDam.Dam_Control_System.service.AiSuggestionService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/api/v1/dam")
@CrossOrigin(origins = "*") // Allow frontend requests to interact during local development
public class DamController {

    private static final Logger log = LoggerFactory.getLogger(DamController.class);

    private final ReservoirStateRepository reservoirStateRepository;
    private final ControlLogRepository controlLogRepository;
    private final DamControlEngineService damControlEngineService;
    private final WeatherForecastService weatherForecastService;
    private final AiSuggestionService aiSuggestionService;
    private final com.SmartDam.Dam_Control_System.repository.EmergencyAlertRepository emergencyAlertRepository;

    public DamController(ReservoirStateRepository reservoirStateRepository,
                         ControlLogRepository controlLogRepository,
                         DamControlEngineService damControlEngineService,
                         WeatherForecastService weatherForecastService,
                         AiSuggestionService aiSuggestionService,
                         com.SmartDam.Dam_Control_System.repository.EmergencyAlertRepository emergencyAlertRepository) {
        this.reservoirStateRepository = reservoirStateRepository;
        this.controlLogRepository = controlLogRepository;
        this.damControlEngineService = damControlEngineService;
        this.weatherForecastService = weatherForecastService;
        this.aiSuggestionService = aiSuggestionService;
        this.emergencyAlertRepository = emergencyAlertRepository;
    }

    /**
     * GET /api/v1/dam/status
     * Returns current dam state and last control decision log for the default dam (Erai).
     */
    @GetMapping("/status")
    public ResponseEntity<Map<String, Object>> getDamStatus() {
        return getDamStatus("erai");
    }

    /**
     * GET /api/v1/dam/{damId}/status
     * Returns current state and last decision log for the specified damId.
     */
    @GetMapping("/{damId}/status")
    public ResponseEntity<Map<String, Object>> getDamStatus(@PathVariable String damId) {
        com.SmartDam.Dam_Control_System.entity.DamMetadata meta = com.SmartDam.Dam_Control_System.entity.DamMetadata.get(damId);
        
        // Retrieve or initialize default Reservoir State if DB is empty for this dam
        ReservoirState state = reservoirStateRepository.findFirstByDamIdOrderByTimestampDesc(meta.getId())
                .orElseGet(() -> {
                    log.info("No reservoir state found in database for dam {}. Initializing default state...", meta.getName());
                    ReservoirState defaultState = ReservoirState.builder()
                            .damId(meta.getId())
                            .timestamp(LocalDateTime.now())
                            .currentVolumeM3(meta.getMaxCapacityM3() * 0.8) // ~80% Capacity
                            .waterLevelMeters(meta.getMaxWaterLevelMeters() * 0.85) // Approximate water level
                            .currentOutflowM3s(0.0)
                            .gateOpenPercentage(0.0)
                            .build();
                    return reservoirStateRepository.save(defaultState);
                });

        ControlLog latestLog = controlLogRepository.findFirstByDamIdOrderByTimestampDesc(meta.getId())
                .orElseGet(() -> {
                    log.info("No control log found in database for dam {}. Initializing default log...", meta.getName());
                    ControlLog defaultLog = ControlLog.builder()
                            .damId(meta.getId())
                            .timestamp(LocalDateTime.now())
                            .forecastPrecipitationMm(0.0)
                            .predictedInflowM3(0.0)
                            .recommendedOutflowM3s(0.0)
                            .floodAlertTriggered(false)
                            .statusMessage("System initialized. Standing by for weather telemetry updates.")
                            .build();
                    return controlLogRepository.save(defaultLog);
                });

        Map<String, Object> response = new HashMap<>();
        response.put("state", state);
        response.put("latestLog", latestLog);
        return ResponseEntity.ok(response);
    }

    /**
     * POST /api/v1/dam/update-state
     */
    @PostMapping("/update-state")
    public ResponseEntity<ControlLog> updateReservoirState(@RequestBody UpdateStateRequest request) {
        return updateReservoirState("erai", request);
    }

    /**
     * POST /api/v1/dam/{damId}/update-state
     * Payload: { "currentVolumeM3": 195000000.0, "waterLevelMeters": 21.5 }
     * Updates reservoir state, fetches forecast rain, runs engine logic, returns decision.
     */
    @PostMapping("/{damId}/update-state")
    public ResponseEntity<ControlLog> updateReservoirState(@PathVariable String damId, @RequestBody UpdateStateRequest request) {
        com.SmartDam.Dam_Control_System.entity.DamMetadata meta = com.SmartDam.Dam_Control_System.entity.DamMetadata.get(damId);
        log.info("Received manual state update request for {}: Volume={}, Level={}",
                meta.getName(), request.getCurrentVolumeM3(), request.getWaterLevelMeters());

        double currentVolume = request.getCurrentVolumeM3();
        double level = request.getWaterLevelMeters();

        // Retrieve or create state record for this dam
        ReservoirState state = reservoirStateRepository.findFirstByDamIdOrderByTimestampDesc(meta.getId())
                .orElse(new ReservoirState());

        state.setDamId(meta.getId());
        state.setTimestamp(LocalDateTime.now());
        state.setCurrentVolumeM3(currentVolume);
        state.setWaterLevelMeters(level);
        reservoirStateRepository.save(state);

        // Fetch weather forecast rain sum
        double forecastRainfall = weatherForecastService.fetchThreeDayPrecipitation(meta.getLatitude(), meta.getLongitude());

        // Run decision engine
        ControlLog updatedDecision = damControlEngineService.evaluateAndExecuteControlLogic(meta.getId(), currentVolume, forecastRainfall);

        return ResponseEntity.ok(updatedDecision);
    }

    /**
     * GET /api/v1/dam/forecast-eval
     */
    @GetMapping("/forecast-eval")
    public ResponseEntity<ControlLog> runForecastSimulation(
            @RequestParam double currentVolumeM3,
            @RequestParam double forecastRainfallMm) {
        return runForecastSimulation("erai", currentVolumeM3, forecastRainfallMm);
    }

    /**
     * GET /api/v1/dam/{damId}/forecast-eval
     * Query Parameters: currentVolumeM3, forecastRainfallMm
     * Triggers on-demand evaluation simulation. Does not write/persist simulation results to DB.
     */
    @GetMapping("/{damId}/forecast-eval")
    public ResponseEntity<ControlLog> runForecastSimulation(
            @PathVariable String damId,
            @RequestParam double currentVolumeM3,
            @RequestParam double forecastRainfallMm) {
        com.SmartDam.Dam_Control_System.entity.DamMetadata meta = com.SmartDam.Dam_Control_System.entity.DamMetadata.get(damId);
        log.info("Received simulation request for {}: Volume={}, Rainfall={}", meta.getName(), currentVolumeM3, forecastRainfallMm);
        ControlLog simulatedLog = damControlEngineService.simulateControlLogic(meta.getId(), currentVolumeM3, forecastRainfallMm);
        return ResponseEntity.ok(simulatedLog);
    }

    // DTO for state update requests
    public static class UpdateStateRequest {
        private double currentVolumeM3;
        private double waterLevelMeters;

        public double getCurrentVolumeM3() {
            return currentVolumeM3;
        }

        public void setCurrentVolumeM3(double currentVolumeM3) {
            this.currentVolumeM3 = currentVolumeM3;
        }

        public double getWaterLevelMeters() {
            return waterLevelMeters;
        }

        public void setWaterLevelMeters(double waterLevelMeters) {
            this.waterLevelMeters = waterLevelMeters;
        }
    }

    /**
     * GET /api/v1/dam/ai-recommendation
     */
    @GetMapping("/ai-recommendation")
    public ResponseEntity<com.SmartDam.Dam_Control_System.dto.AiRecommendationResponse> getAiRecommendation() {
        return getAiRecommendation("erai");
    }

    /**
     * GET /api/v1/dam/{damId}/ai-recommendation
     */
    @GetMapping("/{damId}/ai-recommendation")
    public ResponseEntity<com.SmartDam.Dam_Control_System.dto.AiRecommendationResponse> getAiRecommendation(@PathVariable String damId) {
        log.info("Received request for AI Recommendation for dam: {}", damId);
        com.SmartDam.Dam_Control_System.dto.AiRecommendationResponse recommendation = aiSuggestionService.generateRecommendation(damId);
        return ResponseEntity.ok(recommendation);
    }

    /**
     * POST /api/v1/dam/{damId}/submit-decision
     */
    @PostMapping("/{damId}/submit-decision")
    public ResponseEntity<ControlLog> submitDecision(@PathVariable String damId) {
        log.info("Submitting latest decision log to higher authority for dam: {}", damId);
        com.SmartDam.Dam_Control_System.entity.DamMetadata meta = com.SmartDam.Dam_Control_System.entity.DamMetadata.get(damId);
        
        ControlLog logEntry = controlLogRepository.findFirstByDamIdOrderByTimestampDesc(meta.getId())
                .orElseThrow(() -> new IllegalArgumentException("No log found to submit."));
        
        logEntry.setApprovalStatus("PENDING_AUTHORITY");
        controlLogRepository.save(logEntry);
        return ResponseEntity.ok(logEntry);
    }

    /**
     * POST /api/v1/dam/{damId}/authority-action
     */
    @PostMapping("/{damId}/authority-action")
    public ResponseEntity<ControlLog> authorityAction(
            @PathVariable String damId,
            @RequestParam String action,
            @RequestParam(required = false) Double manualOutflow) {
        log.info("Authority action received for dam: {}. Action: {}, ManualOutflow: {}", damId, action, manualOutflow);
        com.SmartDam.Dam_Control_System.entity.DamMetadata meta = com.SmartDam.Dam_Control_System.entity.DamMetadata.get(damId);
        
        ControlLog logEntry = controlLogRepository.findFirstByDamIdOrderByTimestampDesc(meta.getId())
                .orElseThrow(() -> new IllegalArgumentException("No log found to act on."));

        ReservoirState state = reservoirStateRepository.findFirstByDamIdOrderByTimestampDesc(meta.getId())
                .orElseThrow(() -> new IllegalArgumentException("No reservoir state found."));

        if ("APPROVE".equalsIgnoreCase(action)) {
            logEntry.setApprovalStatus("APPROVED");
            logEntry.setStatusMessage("[APPROVED BY AUTHORITY] " + logEntry.getStatusMessage());
            
            double targetOutflow = logEntry.getRecommendedOutflowM3s();
            state.setCurrentOutflowM3s(targetOutflow);
            
            double openPercent = (targetOutflow / meta.getMaxSafeDischargeM3s()) * 100.0;
            state.setGateOpenPercentage(Math.min(100.0, Math.max(0.0, openPercent)));
            
            reservoirStateRepository.save(state);
        } else {
            logEntry.setApprovalStatus("REJECTED");
            double outflowOverride = (manualOutflow != null) ? manualOutflow : 0.0;
            logEntry.setRecommendedOutflowM3s(outflowOverride);
            logEntry.setStatusMessage(String.format("[OVERRIDDEN BY AUTHORITY] Outflow locked at %.2f m³/s manually.", outflowOverride));
            
            state.setCurrentOutflowM3s(outflowOverride);
            double openPercent = (outflowOverride / meta.getMaxSafeDischargeM3s()) * 100.0;
            state.setGateOpenPercentage(Math.min(100.0, Math.max(0.0, openPercent)));
            
            reservoirStateRepository.save(state);
        }

        controlLogRepository.save(logEntry);
        return ResponseEntity.ok(logEntry);
    }

    /**
     * POST /api/v1/dam/{damId}/emergency-alert
     */
    @PostMapping("/{damId}/emergency-alert")
    public ResponseEntity<com.SmartDam.Dam_Control_System.entity.EmergencyAlert> createEmergencyAlert(
            @PathVariable String damId,
            @RequestParam String priority,
            @RequestParam String message,
            @RequestParam String shiftOfficerName) {
        log.info("Creating emergency alert for dam: {}. Priority: {}, Officer: {}", damId, priority, shiftOfficerName);
        com.SmartDam.Dam_Control_System.entity.EmergencyAlert alert = new com.SmartDam.Dam_Control_System.entity.EmergencyAlert(
                null, damId, priority, message, LocalDateTime.now(), shiftOfficerName, false
        );
        emergencyAlertRepository.save(alert);
        return ResponseEntity.ok(alert);
    }

    /**
     * GET /api/v1/dam/alerts
     */
    @GetMapping("/alerts")
    public ResponseEntity<java.util.List<com.SmartDam.Dam_Control_System.entity.EmergencyAlert>> getActiveAlerts() {
        log.info("Fetching all active/unresolved emergency alerts");
        java.util.List<com.SmartDam.Dam_Control_System.entity.EmergencyAlert> alerts = emergencyAlertRepository.findByResolvedOrderByTimestampDesc(false);
        return ResponseEntity.ok(alerts);
    }

    /**
     * POST /api/v1/dam/alert/{id}/resolve
     */
    @PostMapping("/alert/{id}/resolve")
    public ResponseEntity<Void> resolveAlert(@PathVariable Long id) {
        log.info("Resolving emergency alert ID: {}", id);
        com.SmartDam.Dam_Control_System.entity.EmergencyAlert alert = emergencyAlertRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Alert not found"));
        alert.setResolved(true);
        emergencyAlertRepository.save(alert);
        return ResponseEntity.ok().build();
    }
}

File: src/DamControlSystem/Controllers/DamController.cs
using DamControlSystem.Data.Repositories;
using DamControlSystem.DTOs;
using DamControlSystem.Models;
using DamControlSystem.Services;
using Microsoft.AspNetCore.Mvc;

namespace DamControlSystem.Controllers;

[ApiController]
[Route("api/v1/dam")]
public class DamController : ControllerBase
{
    private readonly ILogger<DamController> _logger;
    private readonly IReservoirStateRepository _reservoirStateRepository;
    private readonly IControlLogRepository _controlLogRepository;
    private readonly IDamControlEngineService _damControlEngineService;
    private readonly IWeatherForecastService _weatherForecastService;
    private readonly IAiSuggestionService _aiSuggestionService;
    private readonly IEmergencyAlertRepository _emergencyAlertRepository;

    public DamController(
        IReservoirStateRepository reservoirStateRepository,
        IControlLogRepository controlLogRepository,
        IDamControlEngineService damControlEngineService,
        IWeatherForecastService weatherForecastService,
        IAiSuggestionService aiSuggestionService,
        IEmergencyAlertRepository emergencyAlertRepository,
        ILogger<DamController> logger)
    {
        _reservoirStateRepository = reservoirStateRepository;
        _controlLogRepository = controlLogRepository;
        _damControlEngineService = damControlEngineService;
        _weatherForecastService = weatherForecastService;
        _aiSuggestionService = aiSuggestionService;
        _emergencyAlertRepository = emergencyAlertRepository;
        _logger = logger;
    }

    /// <summary>
    /// GET /api/v1/dam/status
    /// </summary>
    [HttpGet("status")]
    public Task<IActionResult> GetDamStatus()
    {
        return GetDamStatus("erai");
    }

    /// <summary>
    /// GET /api/v1/dam/{damId}/status
    /// </summary>
    [HttpGet("{damId}/status")]
    public async Task<IActionResult> GetDamStatus(string damId)
    {
        var meta = DamMetadata.Get(damId);

        var state = await _reservoirStateRepository.GetLatestByDamIdAsync(meta.Id);
        if (state == null)
        {
            _logger.LogInformation("No reservoir state found in database for dam {DamName}. Initializing default state...", meta.Name);
            var defaultState = new ReservoirState(
                id: 0,
                damId: meta.Id,
                timestamp: DateTime.UtcNow,
                currentVolumeM3: meta.MaxCapacityM3 * 0.8,
                waterLevelMeters: meta.MaxWaterLevelMeters * 0.85,
                currentOutflowM3s: 0.0,
                gateOpenPercentage: 0.0
            );
            state = await _reservoirStateRepository.SaveAsync(defaultState);
        }

        var latestLog = await _controlLogRepository.GetLatestByDamIdAsync(meta.Id);
        if (latestLog == null)
        {
            _logger.LogInformation("No control log found in database for dam {DamName}. Initializing default log...", meta.Name);
            var defaultLog = new ControlLog(
                id: 0,
                damId: meta.Id,
                timestamp: DateTime.UtcNow,
                forecastPrecipitationMm: 0.0,
                predictedInflowM3: 0.0,
                recommendedOutflowM3s: 0.0,
                floodAlertTriggered: false,
                statusMessage: "System initialized. Standing by for weather telemetry updates."
            );
            latestLog = await _controlLogRepository.SaveAsync(defaultLog);
        }

        return Ok(new
        {
            state,
            latestLog
        });
    }

    /// <summary>
    /// POST /api/v1/dam/update-state
    /// </summary>
    [HttpPost("update-state")]
    public Task<IActionResult> UpdateReservoirState([FromBody] UpdateStateRequest request)
    {
        return UpdateReservoirState("erai", request);
    }

    /// <summary>
    /// POST /api/v1/dam/{damId}/update-state
    /// </summary>
    [HttpPost("{damId}/update-state")]
    public async Task<IActionResult> UpdateReservoirState(string damId, [FromBody] UpdateStateRequest request)
    {
        var meta = DamMetadata.Get(damId);
        _logger.LogInformation("Received manual state update request for {DamName}: Volume={Volume}, Level={Level}",
            meta.Name, request.CurrentVolumeM3, request.WaterLevelMeters);

        var state = await _reservoirStateRepository.GetLatestByDamIdAsync(meta.Id) ?? new ReservoirState();
        state.DamId = meta.Id;
        state.Timestamp = DateTime.UtcNow;
        state.CurrentVolumeM3 = request.CurrentVolumeM3;
        state.WaterLevelMeters = request.WaterLevelMeters;
        await _reservoirStateRepository.SaveAsync(state);

        // Fetch weather forecast rain sum
        double forecastRainfall = await _weatherForecastService.FetchThreeDayPrecipitationAsync(meta.Latitude, meta.Longitude);

        // Run decision engine
        var updatedDecision = await _damControlEngineService.EvaluateAndExecuteControlLogicAsync(
            meta.Id, request.CurrentVolumeM3, forecastRainfall);

        return Ok(updatedDecision);
    }

    /// <summary>
    /// GET /api/v1/dam/forecast-eval
    /// </summary>
    [HttpGet("forecast-eval")]
    public IActionResult RunForecastSimulation([FromQuery] double currentVolumeM3, [FromQuery] double forecastRainfallMm)
    {
        return RunForecastSimulation("erai", currentVolumeM3, forecastRainfallMm);
    }

    /// <summary>
    /// GET /api/v1/dam/{damId}/forecast-eval
    /// </summary>
    [HttpGet("{damId}/forecast-eval")]
    public IActionResult RunForecastSimulation(string damId, [FromQuery] double currentVolumeM3, [FromQuery] double forecastRainfallMm)
    {
        var meta = DamMetadata.Get(damId);
        _logger.LogInformation("Received simulation request for {DamName}: Volume={Volume}, Rainfall={Rainfall}",
            meta.Name, currentVolumeM3, forecastRainfallMm);

        var simulatedLog = _damControlEngineService.SimulateControlLogic(meta.Id, currentVolumeM3, forecastRainfallMm);
        return Ok(simulatedLog);
    }

    /// <summary>
    /// GET /api/v1/dam/ai-recommendation
    /// </summary>
    [HttpGet("ai-recommendation")]
    public Task<IActionResult> GetAiRecommendation()
    {
        return GetAiRecommendation("erai");
    }

    /// <summary>
    /// GET /api/v1/dam/{damId}/ai-recommendation
    /// </summary>
    [HttpGet("{damId}/ai-recommendation")]
    public async Task<IActionResult> GetAiRecommendation(string damId)
    {
        _logger.LogInformation("Received request for AI Recommendation for dam: {DamId}", damId);
        var recommendation = await _aiSuggestionService.GenerateRecommendationAsync(damId);
        return Ok(recommendation);
    }

    /// <summary>
    /// POST /api/v1/dam/{damId}/submit-decision
    /// </summary>
    [HttpPost("{damId}/submit-decision")]
    public async Task<IActionResult> SubmitDecision(string damId)
    {
        _logger.LogInformation("Submitting latest decision log to higher authority for dam: {DamId}", damId);
        var meta = DamMetadata.Get(damId);

        var logEntry = await _controlLogRepository.GetLatestByDamIdAsync(meta.Id);
        if (logEntry == null)
            return BadRequest(new { error = "No log found to submit." });

        logEntry.ApprovalStatus = "PENDING_AUTHORITY";
        await _controlLogRepository.SaveAsync(logEntry);

        return Ok(logEntry);
    }

    /// <summary>
    /// POST /api/v1/dam/{damId}/authority-action
    /// </summary>
    [HttpPost("{damId}/authority-action")]
    public async Task<IActionResult> AuthorityAction(
        string damId,
        [FromQuery] string action,
        [FromQuery] double? manualOutflow = null)
    {
        _logger.LogInformation("Authority action received for dam: {DamId}. Action: {Action}, ManualOutflow: {ManualOutflow}",
            damId, action, manualOutflow);

        var meta = DamMetadata.Get(damId);

        var logEntry = await _controlLogRepository.GetLatestByDamIdAsync(meta.Id);
        if (logEntry == null)
            return BadRequest(new { error = "No log found to act on." });

        var state = await _reservoirStateRepository.GetLatestByDamIdAsync(meta.Id);
        if (state == null)
            return BadRequest(new { error = "No reservoir state found." });

        if ("APPROVE".Equals(action, StringComparison.OrdinalIgnoreCase))
        {
            logEntry.ApprovalStatus = "APPROVED";
            logEntry.StatusMessage = "[APPROVED BY AUTHORITY] " + logEntry.StatusMessage;

            double targetOutflow = logEntry.RecommendedOutflowM3s;
            state.CurrentOutflowM3s = targetOutflow;

            double openPercent = (targetOutflow / meta.MaxSafeDischargeM3s) * 100.0;
            state.GateOpenPercentage = Math.Clamp(openPercent, 0.0, 100.0);

            await _reservoirStateRepository.SaveAsync(state);
        }
        else
        {
            logEntry.ApprovalStatus = "REJECTED";
            double outflowOverride = manualOutflow ?? 0.0;
            logEntry.RecommendedOutflowM3s = outflowOverride;
            logEntry.StatusMessage = $"[OVERRIDDEN BY AUTHORITY] Outflow locked at {outflowOverride:F2} m³/s manually.";

            state.CurrentOutflowM3s = outflowOverride;
            double openPercent = (outflowOverride / meta.MaxSafeDischargeM3s) * 100.0;
            state.GateOpenPercentage = Math.Clamp(openPercent, 0.0, 100.0);

            await _reservoirStateRepository.SaveAsync(state);
        }

        await _controlLogRepository.SaveAsync(logEntry);
        return Ok(logEntry);
    }

    /// <summary>
    /// POST /api/v1/dam/{damId}/emergency-alert
    /// </summary>
    [HttpPost("{damId}/emergency-alert")]
    public async Task<IActionResult> CreateEmergencyAlert(
        string damId,
        [FromQuery] string priority,
        [FromQuery] string message,
        [FromQuery] string shiftOfficerName)
    {
        _logger.LogInformation("Creating emergency alert for dam: {DamId}. Priority: {Priority}, Officer: {Officer}",
            damId, priority, shiftOfficerName);

        var alert = new EmergencyAlert(
            id: 0,
            damId: damId,
            priority: priority,
            message: message,
            timestamp: DateTime.UtcNow,
            shiftOfficerName: shiftOfficerName,
            resolved: false
        );

        var saved = await _emergencyAlertRepository.SaveAsync(alert);
        return Ok(saved);
    }

    /// <summary>
    /// GET /api/v1/dam/alerts
    /// </summary>
    [HttpGet("alerts")]
    public async Task<IActionResult> GetActiveAlerts()
    {
        _logger.LogInformation("Fetching all active/unresolved emergency alerts");
        var alerts = await _emergencyAlertRepository.GetByResolvedAsync(false);
        return Ok(alerts);
    }

    /// <summary>
    /// POST /api/v1/dam/alert/{id}/resolve
    /// </summary>
    [HttpPost("alert/{id:long}/resolve")]
    public async Task<IActionResult> ResolveAlert(long id)
    {
        _logger.LogInformation("Resolving emergency alert ID: {Id}", id);
        var alert = await _emergencyAlertRepository.GetByIdAsync(id);
        if (alert == null)
            return NotFound(new { error = "Alert not found" });

        alert.Resolved = true;
        await _emergencyAlertRepository.SaveAsync(alert);
        return Ok();
    }
}
